### ISSUE TITLE
Multiple ZEO data corruptions due to concurrency bugs

### ISSUE THREAD

**Issue by navytux** [zopefoundation/ZEO#209]
Hello up there. In https://github.com/zopefoundation/ZEO/pull/207 it was [noted](https://github.com/zopefoundation/ZEO/pull/207#issuecomment-1202260612) that current ZEO5 master fails sometimes in `check_race_load_vs_external_invalidate` wrt ZEO4 server. I digged in to find out why and discovered multiple concurrency bugs - in both ZEO4 and ZEO5 - that lead to data corruptions. Please find details about those bugs below:

## Bug1: ZEO5.client - ZEO4.server : race in between load / invalidate

The first bug is that ZEO5.client - contrary to ZEO4.client - does not account for simultaneous invalidations when working wrt ZEO4.server.
It shows as e.g.

```console
(z-dev) kirr@deca:~/src/wendelin/z/ZEO5$ ZEO4_SERVER=1 zope-testrunner -fvvvx --test-path=src -t check_race_load_vs_external
...

AssertionError: T1: obj1 (24)  !=  obj2 (23)
obj1._p_serial: 0x03ea4b6fb05b52cc  obj2._p_serial: 0x03ea4b6faf253855
zconn_at: 0x03ea4b6fb05b52cc  # approximated as max(serials)
zstor.loadBefore(obj1, @zconn.at)       ->  serial: 0x03ea4b6fb05b52cc  next_serial: None
zstor.loadBefore(obj2, @zconn.at)       ->  serial: 0x03ea4b6faf253855  next_serial: 0x03ea4b6fb07def66
zstor._cache.clear()
zstor.loadBefore(obj1, @zconn.at)       ->  serial: 0x03ea4b6fb05b52cc  next_serial: 0x03ea4b6fb07def66
zstor.loadBefore(obj2, @zconn.at)       ->  serial: 0x03ea4b6fb05b52cc  next_serial: 0x03ea4b6fb07def66
```

indicating that obj2 was provided to user from the cache that erroneously got stale.

With [IO trace](https://lab.nexedi.com/kirr/ZEO/commit/f184a1d9) showing message exchange in between client and the server, this looks as:

```
# loadBefore issued
tx (('\x00\x00\x00\x00\x00\x00\x00\x02', '\x03\xeaKo\xaf%8V'), False, 'loadBefore', ('\x00\x00\x00\x00\x00\x00\x00\x02', '\x03\xeaKo\xaf%8V'))

# received invalidateTransaction
rx (0, 1, 'invalidateTransaction', ('\x03\xeaKo\xb0[R\xcc', ['\x00\x00\x00\x00\x00\x00\x00\x01', '\x00\x00\x00\x00\x00\x00\x00\x02']))

# received loadBefore reply but with end_tid=None !!!
rx (('\x00\x00\x00\x00\x00\x00\x00\x02', '\x03\xeaKo\xaf%8V'), 0, '.reply', ('\x80\x03cZODB.tests.MinPO\nMinPO\nq\x01.\x80\x03}q\x02U\x05valueq\x03K\x17s.', '\x03\xeaKo\xaf%8U', None))
```

which:

1) contradicts what @jimfulton [wrote about ZEO4](https://github.com/zopefoundation/ZEO/blob/master/docs/ordering.rst#zeo-4): that there _invalidations are sent in a callback called when the storage lock is held, blocking loads while committing_, and
2) points out what the bug is:

Since in ZEO4 loads can be handled while a commit is in progress, ZEO4.client has special care to detect if an `invalidateTransaction` comes in between `load` request and corresponding `.reply` response, and _does not update the cache if invalidateTransaction sneaked in-between_:

https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/ClientStorage.py#L367-L374
https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/ClientStorage.py#L841-L852
https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/ClientStorage.py#L1473-L1476

but in ZEO5.client `loadBefore` does not have anything like that

https://github.com/zopefoundation/ZEO/blob/fc0729b3cc754bda02c7f54319260b5527dd42a3/src/ZEO/ClientStorage.py#L603-L608
https://github.com/zopefoundation/ZEO/blob/fc0729b3cc754bda02c7f54319260b5527dd42a3/src/ZEO/asyncio/client.py#L289-L309

and thus `invalidateTransaction` sneaked in between `loadBefore` request and
corresponding `.reply` causes ZEO client cache corruption.

In the original `check_race_load_vs_external_invalidate` the problem appears
only sometimes, but the bug happens with ~ 100% probability with the
following delay injected on the server after `loadBefore`:

```diff
--- a/src/ZEO/tests/ZEO4/StorageServer.py
+++ b/src/ZEO/tests/ZEO4/StorageServer.py
@@ -285,7 +285,9 @@ def loadEx(self, oid):
 
     def loadBefore(self, oid, tid):
         self.stats.loads += 1
-        return self.storage.loadBefore(oid, tid)
+        x = self.storage.loadBefore(oid, tid)
+        time.sleep(0.1)
+        return x
 
     def getInvalidations(self, tid):
         invtid, invlist = self.server.get_invalidations(self.storage_id, tid)
```

so maybe, in addition to normal test runs, it could be also good idea to run
the whole ZEO testsuite against so-amended storage backend. This way it will be
similar to how e.g. [races are detected](https://lab.nexedi.com/kirr/go123/blob/070bfdbb/tracing/tracetest/tracetest.go#L76-102) by my [tracetest](https://lab.nexedi.com/kirr/go123/blob/070bfdbb/tracing/tracetest/tracetest.go#L20-73) package.

----

This particular race does not happen for ZEO5.client - ZEO**5**.server, because
there is a difference how invalidations are emitted by the server:

ZEO4 schedules to send invalidations via `callAsyncNoSend`:

https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/StorageServer.py#L1453-L1467

which immediately puts the message into the output queue and arms the trigger to wakeup IO thread:

https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/zrpc/connection.py#L574-L581
https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/zrpc/connection.py#L546-L558

the bug, thus, can happen because `invalidateTransaction` message is put into
the output queue to client immediately on `tpc_finish`.

ZEO5, however, queues to the IO thread both actions via `call_soon_threadsafe`:

https://github.com/zopefoundation/ZEO/blob/fc0729b3cc754bda02c7f54319260b5527dd42a3/src/ZEO/StorageServer.py#L827-L845
https://github.com/zopefoundation/ZEO/blob/fc0729b3cc754bda02c7f54319260b5527dd42a3/src/ZEO/asyncio/server.py#L153-L154

which results in that `invalidateTransaction` message is appended to the output queue at the end of current event loop cycle, not immediately right after `tpc_finish` on the server. In https://github.com/zopefoundation/ZEO/blob/master/docs/ordering.rst#zeo-5 @jimfulton wrote about this the following:

> The server-side hndling of invalidations is a bit tricker in ZEO 5 because there isn't a thread-safe queue of outgoing messages in ZEO 5 as there was in ZEO 4. The natural approach in ZEO 5 would be to use asyncio's `call_soon_threadsafe` to send invalidations in a client's thread. This could easily cause invalidations to be sent after loads. As shown above, this isn't a problem for ZODB 5, at least assuming that invalidations arrive in order. This would be a problem for ZODB 4. For this reason, we require ZODB 5 for ZEO 5.

Anyway, this particular bug does not happen for ZEO5.client - ZEO5.server, because the server does not inject `invalidateTransaction` message in between `loadBefore` request and corresponding `.reply` response.

The fix is either 1) abandon support for working with ZEO4 servers completely, or
2) to implement load-tracking similar to ZEO4 client.

If we go with option "1" I suggest to explicitly reject `Z4` protocol from both ZEO5 client and ZEO5 server.

## Bug2: ZEO5.client - ZEO5.server, ZEO4.client - ZEO4.server : race in between invalidate / disconnect

While investigating Bug1 and verifying whether it also shows itself on ZEO4,
I've found out that, even though ZEO4.client has explicit support for
sneaked-in `invalidateTransaction`, `check_race_load_vs_external_invalidate`
also fails sometimes on ZEO4.client - ZEO4.server, but due to another reason:
failing assertion was similar to the one in Bug1, but IO trace was showing that
for the failure one intermediate `invalidateTransaction` message was not
delivered to client at all. Missed invalidation makes client think that both
objects were unchanged, and retrieve them either from ZODB.Connection, or ZEO
cache, but if for some reason, one of the object is evicted from the cache, it
will result in inconsistency detected by broken test invariant.

The bug here is due to the race on the server: when sending out
`invalidateTransaction` messages to all connected clients, both ZEO4 and ZEO5
iterate the list of connected clients without any locking:

https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/StorageServer.py#L1108-L1115

https://github.com/zopefoundation/ZEO/blob/fc0729b3cc754bda02c7f54319260b5527dd42a3/src/ZEO/StorageServer.py#L843-L845

So, if during such iteration, a client disconnects, it will cause some _other_
client to be skipped during the iteration and so do not receive
currently-processed `invalidateTransaction` message. The following code
illustrates this:

```
In [9]: l = [1,2,3,4]		# list of clients

In [10]: i = iter(l)		# start iterating over it

In [11]: next(i)		# first client yielded
Out[11]: 1

In [12]: next(i)		# second client yielded
Out[12]: 2

In [13]: del l[0]		# first client disconnects

In [14]: l
Out[14]: [2, 3, 4]

In [15]: next(i)		# `3` is skipped wrongly
Out[15]: 4

In [16]: next(i)
---------------------------------------------------------------------------
StopIteration                             Traceback (most recent call last)
```

This bug affects both servers: ZEO4 and ZEO5 in multi-threaded mode. ZEO5 is
affected only in multithreaded mode, because if ZEO5 server runs
single-threaded, a disconnection should not be handled simultaneously to
`invalidateTransaction` processing. ZEO5 multi-threaded mode is activated by
`ZEO_MTACCEPTOR=1` environment variable. This mode is scheduled to be removed
in ZEO6, but still, my understanding is that the bug does not show itself in
single-threaded mode only due to chance, not by design - there is neither
comments, nor explicit care to avoid this race.

The fix is to iterate over the snapshot copy of the list of clients when doing
such kind of iterations.

**Comment by navytux** [comment]
Tests for Bug2 : https://github.com/zopefoundation/ZODB/pull/368.

**Comment by navytux** [comment]
Fix for Bug2 in ZEO4: https://lab.nexedi.com/kirr/ZEO/commit/b71e76ae.

**Comment by d-maurer** [comment]
Kirill Smelkov wrote at 2022-8-15 11:34 -0700:
> ...
>While investigating Bug1 and verifying whether it also shows itself on ZEO4,
>I've found out that, even though ZEO4.client has explicit support for
>sneaked-in `invalidateTransaction`, `check_race_load_vs_external_invalidate`
>also fails sometimes on ZEO4.client - ZEO4.server, but due to another reason:
>failing assertion was similar to the one in Bug1, but IO trace was showing that
>for the failure one intermediate `invalidateTransaction` message was not
>delivered to client at all. Missed invalidation makes client think that both
>objects were unchanged, and retrieve them either from ZODB.Connection, or ZEO
>cache, but if for some reason, one of the object is evicted from the cache, it
>will result in inconsistency detected by broken test invariant.
>
>The bug here is due to the race on the server: when sending out
>`invalidateTransaction` messages to all connected clients, both ZEO4 and ZEO5
>iterate the list of connected clients without any locking:

At least for ZEO5, the invalidation message sending should happen
in a coroutine context. This context should not get disturbed
by a disconnection (noticed only after the coroutine suspends itself).

**Comment by navytux** [comment]
@d-maurer, thanks for feedback. Invalidation sending happens in coroutine context in the IO thread - yes, but if the storage is multithreaded - the acceptor code in another thread can mutate the data structure, that IO thread is using, simultaeneously. And that leads to some clients being skipped on sending the invalidation message.

Here I'm talking about `ZEO_MTACCEPTOR=1` mode. Yes, this mode recently became deprecated, but there was no conscious understanding that multithreading mode removal is needed for correctness. If this is indeed the case, I believe we should explicitly document single-threaded assumption in the places where it is important.

**Comment by d-maurer** [comment]
Kirill Smelkov wrote at 2022-8-15 23:39 -0700:
***@***.***, thanks for feedback. Invalidation sending happens in coroutine context in the IO thread - yes, but if the storage is multithreaded - the acceptor code in another thread can mutate the data structure, that IO thread is using, simultaeneously.

I think that use case is already scheduled to be dropped.
When I remember right, tests for it have already been abandonned.

Significant complexity has been introduced to the server code
to support this use case. In particular, the use of locks
depends on whether an `Acceptor` (storage server essentially
run by a single thread) or `MTAcceptor` (each connection
has its own thread) is used.
Dropping support for it can lead to significant simplifications.


>Here I'm talking about `ZEO_MTACCEPTOR=1` mode. Yes, this mode recently became deprecated, but there was no conscious understanding that multithreading mode removal is needed for correctness. If this is indeed the case, I believe we should explicitly document single-threaded assumption in the places where it is important.

Apparently, nobody monitoring the ZEO related discussions on `github`
is using it. Otherwise, I would have expected a reaction when
@dataflake asked whether it can be abandoned.

The difficulties you have had to reproduce the bug
indicates that even if somebody is using the scenario
he likely will not get affected by the error: it is quite
unusual that an application frequently opens new storage connections.

Thus, while this is an incorrectness, it likely does not require
special actions.
I have the feeling that it may be sufficient to officially drop
support for this use case -- as already planned.

**Comment by dataflake** [comment]
I agree with Dieter. I don't see any value in a wild goose chase uncovering issues with a non-standard implementation that hasn't seen much use and zero active support for many years.

**Comment by d-maurer** [comment]
Kirill Smelkov wrote at 2022-8-15 11:34 -0700:
>Hello up there. In https://github.com/zopefoundation/ZEO/pull/207 it was [noted](https://github.com/zopefoundation/ZEO/pull/207#issuecomment-1202260612) that current ZEO5 master fails sometimes in `check_race_load_vs_external_invalidate` wrt ZEO4 server. I digged in to find out why and discovered multiple concurrency bugs

Great!

**Comment by navytux** [comment]
@d-maurer, @dataflake, thanks for feedback, kind words, and I apologize for the delay with replying on my side.

@dataflake, please do not get me wrong. In the first place I'm not trying to raise up incorrectness issues in ancient configurations, that noone is actually using. What we had in https://github.com/zopefoundation/ZEO/pull/207#issuecomment-1202260612 was sign of data corruption due to a concurrency bug, and it was unclear where the bug sits and whether it affects only old code we stopped (or are stopping) to care about, or modern configuration as well. A data corruption is potentially significant problem in my view. It's like when an aircraft or spaceship crashes, or shows significant issues, all other aircrafts of the same model are landed and stop to fly until the issue is understood. Isn't it?

Said that, with current understanding, I suggest the following steps:

1. remove support for ZEO4 from ZEO5 client without waiting for ZEO6. The data corruption described by Bug1 is significant in my view and deserves the fix without waiting, nor without forcing users to upgrade to new major version. If we keep support for ZEO4 server, the fix is non-trivial. But in situation, when, I believe, nobody is using ZEO5.client - ZEO5.server configuration today, the fix could be to reject connection to ZEO4 server at all. If upon such rejection we log appropriate message, explaining that ZEO4 server support is dropped because of the data corruption reasons, I believe it should be clear for people, even if someone is using such configuration, to stop using it. And we can also add to the logged message, that if, for some reason, someone actually depends on such configuration today, that the person should speak up and describe those reasons on a ZEO issue. After all in 2015/2016 it was practical to add ZEO4 compatibility to newly-introduced and unstable ZEO5 client. But in 2022 the situation is different - ZEO5 is believed to be relatively stable, and ZEO4 stopped to receive updates several years ago. That's why I believe that dropping ZEO4 server support from ZEO5 should be ok.


2. Regarding MTACCEPTOR in ZEO5 server, we already scheduled removal of this mode in ZEO 5.3.0 with the following entry in the changelog:

   > Remove tests for the asyncio/mtacceptor module, it appears unused and presents a maintenance burden. The module will be removed in ZEO version 6.

   (https://pypi.org/project/ZEO/)

   and now, given that this mode also exposes data corruption issues I would
   also remove it with corresponding changelog entry without waiting for ZEO6 to
   be released.


What do you think?

If we agree on 1 and/or 2, I think I could implement those steps.

Kirill

**Comment by dataflake** [comment]
The general policy throughout Zope Foundation projects for feature removals is this: Feature removal is normally preceded by clear deprecation warnings in the current release branch and the final removal requires a new major release. That way nasty surprises for users are kept to a minimum.

I can agree with removing MTACCEPTOR earlier - I think it's truly unused. Scheduling it for ZEO 6 was just an expression of our feature removal policy.

As you know my main concern (or main use case) for ZODB and ZEO is Zope deployments. So I checked the ZEO version required by Zope 4 and Zope 5 and noticed that ZEO 5 has been required from the earliest releases of Zope 4. That means there's a super low probability of Zope 4/5 deployments running on ZEO 4. I also noticed that ZEO 4 has last been released as version 4.2.1 6 years ago, it's basically dead. So I can agree with removal of ZEO 4 support as well.

In summary, both 1 and 2 are fine by me. I see very little risk breaking anyone's Zope deployment. And if you believe the data corruption risk is significant that's a good reason.

**Comment by navytux** [comment]
@dataflake, thanks for feedback, and it is good to see we generally agree here. I've created https://github.com/zopefoundation/ZEO/pull/211 to proceed with 2. Could you please have a look?

**Comment by navytux** [comment]
I've created https://github.com/zopefoundation/ZEO/pull/212 to start work on 1.

**Comment by navytux** [comment]
Pull request to remove support for interoperability with ZEO4 server: https://github.com/zopefoundation/ZEO/pull/213.

**Comment by navytux** [comment]
Fixed by https://github.com/zopefoundation/ZEO/pull/211 and https://github.com/zopefoundation/ZEO/pull/213.
